src/System.Linq.Parallel/tests/ExchangeTests.cs

```diff
@@ -10,183 +11,177 @@ namespace Test
+        public static void Partitioning_Striped(Labeled<ParallelQuery<int>> labeled, int count, int partitions)
         {
-            int[] data = new int[dataSize];
-            for (int i = 0; i < dataSize; i++)
-                data[i] = i;
-
-            var whereOp = data.AsParallel().Where(
-                delegate (int x) { return (x % 2) == 0; }); // select only even elements
-            var selectOp = whereOp.Select(
-                delegate (int x) { return x * 2; }); // just double the elements
-
-            // Now verify the output is what we expect.
-
-            int expectSum = 0;
-            for (int i = 0; i < dataSize; i++)
-                if ((i % 2) == 0)
-                    expectSum += (i * 2);
-
-            int realSum = 0;
-            IEnumerator<int> e = selectOp.GetEnumerator();
-            while (e.MoveNext())
+            int seen = 0;
+            foreach (int i in labeled.Item.WithDegreeOfParallelism(partitions).Take(count).Select(i => i))
             {
-                realSum += e.Current;
+                Assert.Equal(seen++, i);
             }
+        }
+
+        [Theory]
+        [OuterLoop]
+        [MemberData("PartitioningData", (object)(new int[] { 1024 * 4, 1024 * 1024 }))]
+        public static void Partitioning_Striped_Longrunning(Labeled<ParallelQuery<int>> labeled, int count, int partitions)
+        {
+            Partitioning_Striped(labeled, count, partitions);
+        }
 
-            if (realSum != expectSum)
+        [Theory]
+        [MemberData("MergeData", (object)(new int[] { 0, 1, 2, 16, 1024 }))]
+        public static void Merge_Ordered(Labeled<ParallelQuery<int>> labeled, int count, ParallelMergeOptions options)
+        {
+            int seen = 0;
+            foreach (int i in labeled.Item.WithMergeOptions(options).Select(i => i))
             {
-                Assert.True(false, string.Format("CheckWhereSelectComposition datasize({0}):  actual sum does not equal expected sum.", dataSize));
+                Assert.Equal(seen++, i);
             }
         }
 
-        [Fact]
-        public static void PartitioningTest()
+        [Theory]
+        [OuterLoop]
+        [MemberData("MergeData", (object)(new int[] { 1024 * 4, 1024 * 1024 }))]
+        public static void Merge_Ordered_Longrunning(Labeled<ParallelQuery<int>> labeled, int count, ParallelMergeOptions options)
         {
-            PartitioningTestCore(true, 1, 0, 10);
-            PartitioningTestCore(false, 1, 0, 10);
-
-            PartitioningTestCore(true, 1, 0, 500);
-            PartitioningTestCore(false, 1, 0, 520);
+            Merge_Ordered(labeled, count, options);
+        }
 
-            PartitioningTestCore(true, 4, 0, 500);
-            PartitioningTestCore(false, 4, 0, 520);
+        [Theory]
+        [MemberData("MergeData", (object)(new int[] { 16, 1024 }))]
+        [MemberData("FailingMergeData", (object)(new int[] { 16, 1024 }))]
```

Comment:
I'm confused by the naming.  What is "FailingMergeData"?  I'd have expected to see some Asserts.Throws or try/catches or something in the test, as I assumed from the name that meant something was going to fail and throw.